File: dashboard_bonsai/src/kv_row.ml
(** KvRow — atomic label/value row primitive (Bonsai mirror of Preact
    [kv-row.ts]).

    See [kv_row.mli] for the public contract.

    SPEC mapping (primitives.css [.kv-row]):
    - grid-template-columns: 80px 1fr  ([.is-wide] → 120px 1fr)
    - gap: var(--sp-3) (12px)
    - padding: 4px 0
    - align-items: baseline
    - font-size: 11px
    - .k → fg-muted, fs-10, uppercase, letter-spacing 0.06em
    - .v → fg-primary, mono, fs-11

    Mirrors the Preact [MONO_STACK] literal so the bonsai island
    renders the same monospace fallback chain as the Preact runtime. *)

open! Core
open! Bonsai_web
open Virtual_dom.Vdom

module Style =
[%css
stylesheet
  {|
  .row {
    display: grid;
    grid-template-columns: 80px 1fr;
    gap: 12px;
    padding: 4px 0;
    align-items: baseline;
    font-size: var(--fs-11, 11px);
  }

  .row_wide {
    grid-template-columns: 120px 1fr;
  }

  .k {
    color: var(--color-fg-muted);
    font-size: 10px;
    text-transform: uppercase;
    letter-spacing: 0.06em;
  }

  .v {
    color: var(--color-fg-primary);
    font-family: ui-monospace, SFMono-Regular, "SF Mono", Menlo, Consolas, "Liberation Mono", monospace;
    font-size: var(--fs-11, 11px);
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  .v_wrap {
    white-space: normal;
    overflow: visible;
    text-overflow: clip;
    word-break: break-all;
  }
|}]

type width =
  [ `Default
  | `Wide
  ]

let width_class : width -> Attr.t list = function
  | `Default -> []
  | `Wide -> [ Style.row_wide ]
;;

let width_data_value : width -> string = function
  | `Default -> "false"
  | `Wide -> "true"
;;

let view
      ?(width = `Default)
      ?(wrap = false)
      ?value
      ?(children = [])
      ?testid
      ~label
      ()
  : Node.t
  =
  let row_attrs =
    let base =
      Style.row
      :: width_class width
      @ [ Attr.create "data-kv-row" ""
        ; Attr.create "data-kv-wide" (width_data_value width)
        ]
    in
    match testid with
    | Some id -> Attr.create "data-testid" id :: base
    | None -> base
  in
  let key =
    Node.span
      ~attrs:[ Style.k; Attr.create "data-kv-key" "" ]
      [ Node.text label ]
  in
  let value_node =
    match children with
    | _ :: _ -> Node.div children
    | [] ->
      let v_attrs =
        let base = [ Style.v; Attr.create "data-kv-value" "" ] in
        if wrap then Style.v_wrap :: base else base
      in
      Node.span ~attrs:v_attrs [ Node.text (Option.value value ~default:"") ]
  in
  Node.div ~attrs:row_attrs [ key; value_node ]
;;

File: dashboard/src/components/kv-row.ts
// KvRow — atomic primitive ported from design-system v0.4
// primitives.html (`<div class="kv-row"><span class="k">LABEL</span>
// <span class="v">value</span></div>`). The SPEC defines a single
// label / value row with a fixed 80px label column, 4px vertical
// padding, baseline alignment, uppercase muted label, and a mono
// value in primary fg. Used inside drawer details, inspector panels,
// keeper-detail KV strips — anywhere a screen says "here are the
// metadata facts about this thing".
//
// Distinct from existing dashboard primitives:
//
//   AuthRow (auth-status.ts) — `display:contents` cells that fill a
//     parent grid. Coupled to its parent's grid template; not a
//     self-contained row. This PR replaces all four AuthRow callers
//     with KvRow + a `flex flex-col` parent.
//   StatCell / KpiCell — value-first metric tiles, not label/value
//     symmetry. Different intent.
//
// SPEC mapping (primitives.css `.kv-row`):
//   grid-template-columns: 80px 1fr  (`is-wide` → 120px 1fr)
//   gap: var(--sp-3) (12px)
//   padding: 4px 0
//   align-items: baseline
//   font-size: 11px
//   .k → fg-muted, fs-10, uppercase, letter-spacing 0.06em
//   .v → fg-primary, mono, fs-11

import { html } from 'htm/preact'
import type { ComponentChildren, VNode } from 'preact'
import { MONO_STACK } from './common/font-stacks'

export interface KvRowProps {
  /** Label text (rendered uppercase via the SPEC font rule). */
  label: string
  /** Value content. Either a string (rendered as the SPEC mono span)
   *  or arbitrary children (chips, pills, links — caller controls
   *  font then). When children are provided, `value` is ignored. */
  value?: string
  children?: ComponentChildren
  /** SPEC `.kv-row.is-wide` — 120px label column. Default false. */
  wide?: boolean
  /** Allow long values to wrap on word boundaries. SPEC value is
   *  whitespace:nowrap by default; long ids / paths often need wrap.
   *  Default false (matches SPEC). */
  wrap?: boolean
  /** Forwarded to data-testid on the row container. */
  testId?: string
}


export function KvRow(props: KvRowProps): VNode {
  const wide = props.wide === true

  const rowStyle = {
    display: 'grid',
    gridTemplateColumns: wide ? '120px 1fr' : '80px 1fr',
    gap: 'var(--sp-3)',
    padding: 'var(--sp-1) 0',
    alignItems: 'baseline' as const,
    fontSize: 'var(--fs-11)',
  }

  const keyStyle = {
    color: 'var(--color-fg-muted)',
    fontSize: 'var(--fs-10)',
    textTransform: 'uppercase' as const,
    letterSpacing: '0.06em',
  }

  const valueStyle = {
    color: 'var(--color-fg-primary)',
    fontFamily: MONO_STACK,
    fontSize: 'var(--fs-11)',
    whiteSpace: props.wrap === true ? 'normal' as const : 'nowrap' as const,
    overflow: props.wrap === true ? undefined : 'hidden',
    textOverflow: props.wrap === true ? undefined : 'ellipsis',
    wordBreak: props.wrap === true ? 'break-all' as const : undefined,
  }

  // Children win over the value string — callers passing chips /
  // pills as the value need their own typography, not the mono span.
  const valueContent = props.children != null
    ? props.children
    : html`<span data-kv-value style=${valueStyle}>${props.value ?? ''}</span>`

  return html`
    <div
      data-testid=${props.testId}
      data-kv-row
      data-kv-wide=${wide ? 'true' : 'false'}
      style=${rowStyle}
    >
      <span data-kv-key style=${keyStyle}>${props.label}</span>
      ${valueContent}
    </div>
  `
}
